syndesis-maven-plugin/src/main/java/io/syndesis/maven/ExtractConnectorDescriptorsMojo.java

```diff
@@ -0,0 +1,284 @@
+import org.apache.maven.shared.artifact.resolve.ArtifactResolver;
+import org.apache.maven.shared.artifact.resolve.ArtifactResolverException;
+import org.apache.maven.shared.artifact.resolve.ArtifactResult;
+
+/**
+ * Simple maven plugin for creating a JSON file holding all connectors descriptors which are supported.
+ *
+ * @author roland
+ * @since 23.09.17
+ */
+@Mojo(name = "extract-connector-descriptors", defaultPhase = LifecyclePhase.GENERATE_RESOURCES)
+public class ExtractConnectorDescriptorsMojo extends AbstractMojo {
+
+    @Component
+    private ArtifactResolver artifactResolver;
+
+    @Parameter(defaultValue = "${session}", readonly = true, required = true)
+    private MavenSession session;
+
+    @Parameter(defaultValue = "${project.remoteArtifactRepositories}", readonly = true, required = true)
+    private List<ArtifactRepository> remoteRepositories;
+
+    @Parameter(defaultValue = "${project.build.outputDirectory}/META-INF/camel/camel-meta.json")
+    private String target;
+
+    @Override
+    @SuppressWarnings("PMD.EmptyCatchBlock")
+    public void execute() throws MojoExecutionException, MojoFailureException {
+
+        ArrayNode root = new ArrayNode(JsonNodeFactory.instance);
+
+        URLClassLoader classLoader = null;
+        try {
+            PluginDescriptor desc = (PluginDescriptor) getPluginContext().get("pluginDescriptor");
+            List<Artifact> artifacts = desc.getArtifacts();
+            ProjectBuildingRequest buildingRequest =
+                new DefaultProjectBuildingRequest(session.getProjectBuildingRequest());
+            buildingRequest.setRemoteRepositories(remoteRepositories);
+            for (Artifact artifact : artifacts) {
+                ArtifactResult result = artifactResolver.resolveArtifact(buildingRequest, artifact);
+                File jar = result.getArtifact().getFile();
+                classLoader = createClassLoader(jar);
+                if (classLoader == null) {
+                    throw new IOException("Can not create classloader for " + jar);
+                }
+                ObjectNode entry = new ObjectNode(JsonNodeFactory.instance);
+                addConnectorMeta(entry, classLoader);
+                    addComponentMeta(entry, classLoader);
+                if (entry.size() > 0) {
+                    addGav(entry, artifact);
+                    root.add(entry);
+                }
+            }
+            if (root.size() > 0) {
+                saveCamelMetaData(root);
+            }
+        } catch (ArtifactResolverException | IOException e) {
+            throw new MojoExecutionException(e.getMessage(), e);
+        } finally {
+            if (classLoader != null) {
+                try {
+                    classLoader.close();
+                } catch (IOException ignored) {
+
+                }
+            }
+        }
+    }
+
+    private URLClassLoader createClassLoader(File jar) throws MalformedURLException {
```

Comment:
I wonder if we need a classloader or can use just `JarInputStream`, we're just loading JSON/properties files not classes right?

Reply:
Yes, just tried to use a JarInputStream. But reading real data from a JarInputStream is quite tedious (without proper helper which I couldn't find any). So would like to keep that. I like the simplicity of getResourceAsStream.